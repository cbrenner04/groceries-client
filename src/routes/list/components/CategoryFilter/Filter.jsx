import React from 'react';
import PropTypes from 'prop-types';
import { Dropdown } from 'react-bootstrap';

const Filter = ({ categories, handleCategoryFilter }) => (
  <Dropdown data-test-id="filter-dropdown">
    <Dropdown.Toggle variant="light" id="filter-by-category-button">
      Filter by category
    </Dropdown.Toggle>
    <Dropdown.Menu>
      {categories.sort().map((category) => {
        if (!category) return '';
        return (
          <Dropdown.Item as="button" key={category} name={category} onClick={handleCategoryFilter}>
            {category}
          </Dropdown.Item>
        );
      })}
    </Dropdown.Menu>
  </Dropdown>
);

Filter.propTypes = {
  categories: PropTypes.arrayOf(PropTypes.string).isRequired,
  handleCategoryFilter: PropTypes.func.isRequired,
};

export default Filter;
