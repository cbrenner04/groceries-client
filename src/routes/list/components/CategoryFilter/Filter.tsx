import React, { MouseEventHandler } from 'react';
import { Dropdown } from 'react-bootstrap';

interface IFilterProps {
  categories: string[];
  handleCategoryFilter: MouseEventHandler;
}

const Filter: React.FC<IFilterProps> = ({ categories, handleCategoryFilter }) => (
  <Dropdown data-test-id="filter-dropdown">
    <Dropdown.Toggle variant="light" id="filter-by-category-button">
      Filter by category
    </Dropdown.Toggle>
    <Dropdown.Menu>
      {categories.sort().map((category) => {
        if (!category) {
          return '';
        }
        return (
          <Dropdown.Item
            as="button"
            key={category}
            name={category}
            onClick={handleCategoryFilter}
            data-test-id={`filter-by-${category}`}
          >
            {category}
          </Dropdown.Item>
        );
      })}
    </Dropdown.Menu>
  </Dropdown>
);

export default Filter;
