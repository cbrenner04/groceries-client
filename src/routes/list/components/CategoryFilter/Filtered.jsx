import React from 'react';
import PropTypes from 'prop-types';
import { Button } from 'react-bootstrap';

const Filtered = ({ filter, handleClearFilter }) => (
  <>
    <span id="filter-title">Filtering by:</span>
    <Button data-test-id="clear-filter" variant="outline-primary" id="filter-button" onClick={handleClearFilter}>
      {filter} <i className="fa fa-trash" />
    </Button>
  </>
);

Filtered.propTypes = {
  filter: PropTypes.string.isRequired,
  handleClearFilter: PropTypes.func.isRequired,
};

export default Filtered;
