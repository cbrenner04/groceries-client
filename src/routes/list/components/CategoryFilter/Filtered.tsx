import React, { type MouseEventHandler } from 'react';
import { Button } from 'react-bootstrap';

export interface IFilteredProps {
  filter: string;
  handleClearFilter: MouseEventHandler;
}

const Filtered: React.FC<IFilteredProps> = (props): React.JSX.Element => (
  <React.Fragment>
    <span id="filter-title">Filtering by:</span>
    <Button data-test-id="clear-filter" variant="outline-primary" id="filter-button" onClick={props.handleClearFilter}>
      {props.filter} <i className="fa fa-trash" />
    </Button>
  </React.Fragment>
);

export default Filtered;
