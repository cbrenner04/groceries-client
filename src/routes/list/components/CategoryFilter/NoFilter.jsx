import React from 'react';
import { Button } from 'react-bootstrap';

const NoFilter = () => (
  <Button variant="light" disabled style={{ cursor: 'not-allowed' }}>
    Filter by category
  </Button>
);

export default NoFilter;
