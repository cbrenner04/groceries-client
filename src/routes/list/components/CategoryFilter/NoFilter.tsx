import React from 'react';
import { Button } from 'react-bootstrap';

const NoFilter: React.FC = () => (
  <Button variant="light" id="no-filter" disabled>
    Filter by category
  </Button>
);

export default NoFilter;
