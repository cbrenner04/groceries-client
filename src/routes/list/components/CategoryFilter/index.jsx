import React from 'react';
import PropTypes from 'prop-types';

import Filter from './Filter';
import Filtered from './Filtered';
import NoFilter from './NoFilter';

const CategoryFilter = (props) => {
  let component;
  if (props.categories.filter(Boolean).length) {
    if (props.filter) {
      component = <Filtered filter={props.filter} handleClearFilter={props.handleClearFilter} />;
    } else {
      component = <Filter categories={props.categories} handleCategoryFilter={props.handleCategoryFilter} />;
    }
  } else {
    component = <NoFilter />;
  }

  return <div className="float-right">{component}</div>;
};

CategoryFilter.propTypes = {
  categories: PropTypes.arrayOf(PropTypes.string),
  filter: PropTypes.string,
  handleClearFilter: PropTypes.func.isRequired,
  handleCategoryFilter: PropTypes.func.isRequired,
};

CategoryFilter.defaultProps = {
  filter: '',
  categories: [''],
};

export default CategoryFilter;
