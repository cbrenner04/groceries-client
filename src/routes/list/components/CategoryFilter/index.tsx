import React from 'react';
import PropTypes from 'prop-types';

import Filter from './Filter';
import Filtered from './Filtered';
import NoFilter from './NoFilter';

const CategoryFilter = (props) => {
  const cats = props.categories || [''];
  if (cats.filter(Boolean).length) {
    if (props.filter || '') {
      return <Filtered filter={props.filter || ''} handleClearFilter={props.handleClearFilter} />;
    } else {
      return <Filter categories={cats} handleCategoryFilter={props.handleCategoryFilter} />;
    }
  } else {
    return <NoFilter />;
  }
};

CategoryFilter.propTypes = {
  categories: PropTypes.arrayOf(PropTypes.string),
  filter: PropTypes.string,
  handleClearFilter: PropTypes.func.isRequired,
  handleCategoryFilter: PropTypes.func.isRequired,
};

export default CategoryFilter;
