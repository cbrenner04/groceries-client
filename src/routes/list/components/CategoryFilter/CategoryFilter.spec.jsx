import React from 'react';
import { render } from '@testing-library/react';

import CategoryFilter from './index';

describe('CategoryFilter', () => {
  it('renders Filtered when categories and filter exist', () => {
    const props = {
      categories: ['foo', 'bar'],
      filter: 'foo',
      handleClearFilter: jest.fn(),
      handleCategoryFilter: jest.fn(),
    };
    const { container, getByRole } = render(<CategoryFilter {...props} />);

    expect(container).toMatchSnapshot();
    expect(getByRole('button')).toHaveTextContent('foo');
  });

  it('renders Filter when categories exist but filter does not', () => {
    const props = {
      categories: ['foo', 'bar'],
      filter: '',
      handleClearFilter: jest.fn(),
      handleCategoryFilter: jest.fn(),
    };
    const { container, getByRole } = render(<CategoryFilter {...props} />);

    expect(container).toMatchSnapshot();
    expect(getByRole('button')).toHaveTextContent('Filter by category');
    expect(getByRole('button')).not.toBeDisabled();
  });

  it('renders NoFilter when categories do not exist', () => {
    const props = {
      categories: [],
      filter: '',
      handleClearFilter: jest.fn(),
      handleCategoryFilter: jest.fn(),
    };
    const { container, getByRole } = render(<CategoryFilter {...props} />);

    expect(container).toMatchSnapshot();
    expect(getByRole('button')).toHaveTextContent('Filter by category');
    expect(getByRole('button')).toBeDisabled();
  });
});
