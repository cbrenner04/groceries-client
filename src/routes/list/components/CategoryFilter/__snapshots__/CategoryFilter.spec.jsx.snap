// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryFilter renders Filter when categories exist but filter does not 1`] = `
<div>
  <div
    class="dropdown"
    data-test-id="filter-dropdown"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="dropdown-toggle btn btn-light"
      id="filter-by-category-button"
      type="button"
    >
      Filter by category
    </button>
  </div>
</div>
`;

exports[`CategoryFilter renders Filtered when categories and filter exist 1`] = `
<div>
  <span
    id="filter-title"
  >
    Filtering by:
  </span>
  <button
    class="btn btn-outline-primary"
    data-test-id="clear-filter"
    id="filter-button"
    type="button"
  >
    foo
     
    <i
      class="fa fa-trash"
    />
  </button>
</div>
`;

exports[`CategoryFilter renders NoFilter when categories do not exist 1`] = `
<div>
  <button
    class="btn btn-light"
    disabled=""
    id="no-filter"
    type="button"
  >
    Filter by category
  </button>
</div>
`;
